### PR TITLE
Don't Auto-show UnknownDeviceDialog

### PR DESCRIPTION
**Pull request by dbkr**
- [x] Make the dialog not pop-up
 - [x] Switch to using js-sdk native events to watch for pending unsent events
 - [x] Make VoIP calls work sensibly with UnknownDeviceErrors (ie. not play ringback tones while the events haven't been sent)
 - [x] Make the dialog look more sensible if you actually verify all the devices (maybe 'send anyway' just turns into 'send')

This has turned into a grab-bag of mostly VoIP fixes (see js-sdk PR) but also hopefully make the UnknownDeviceDialog a bit friendlier and more obvious.

Basically this works by listening for the pending events changing and updating the state of the bar based on that, rather than dispatching separately.

Requires https://github.com/matrix-org/matrix-js-sdk/pull/575

### CODE REVIEW

File: src/CallHandler.js [hunk 1]
```diff
@@ -97,19 +99,54 @@ function pause(audioId) {
...
+    } else {
+        call.answer();
+    }
+}
+
 function _setCallListeners(call) {
     call.on("error", function(err) {
         console.error("Call error: %s", err);
         console.error(err.stack);
-        call.hangup();
```
Comment (richvdh) [inline src/CallHandler.js]: do we not still need this?

File: src/CallHandler.js [hunk 2]
```diff
@@ -97,19 +99,54 @@ function pause(audioId) {
...
-        _setCallState(undefined, call.roomId, "ended");
-    });
-    call.on('send_event_error', function(err) {
```
Comment (richvdh) [inline src/CallHandler.js]: and this?

File: src/CallHandler.js [hunk 3]
```diff
@@ -179,7 +216,6 @@ function _setCallState(call, roomId, status) {
 function _onAction(payload) {
     function placeCall(newCall) {
         _setCallListeners(newCall);
-        _setCallState(newCall, newCall.roomId, "ringback");
```
Comment (richvdh) [inline src/CallHandler.js]: and this?

File: src/components/structures/RoomStatusBar.js [hunk 4]
```diff
@@ -136,6 +147,26 @@ module.exports = React.createClass({
...
+
+    _onCancelAllClick: function() {
+        Resend.cancelUnsentEvents(this.props.room);
+    },
+
+    _onShowDevicesClick: function() {
+        showUnknownDeviceDialogForMessages(MatrixClientPeg.get(), this.props.room);
+    },
+
+    onRoomLocalEchoUpdated: function(event, room, oldEventId, oldStatus) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/RoomStatusBar.js]: done

File: src/components/structures/RoomStatusBar.js [hunk 5]
```diff
@@ -241,6 +272,61 @@ module.exports = React.createClass({
         return avatars;
     },
 
+    _getUnsentMessageContent: function(room) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/RoomStatusBar.js]: done

File: src/components/views/dialogs/UnknownDeviceDialog.js [hunk 6]
```diff
@@ -83,36 +93,69 @@ function UnknownDeviceList(props) {
...
-            Object.keys(this.props.devices[userId]).map((deviceId) => {
-                MatrixClientPeg.get().setDeviceKnown(userId, deviceId, true);
-            });
-        });
+    componentWillMount: function() {
+        MatrixClientPeg.get().on("deviceVerificationChanged", this._onDeviceVerificationChanged);
+    },
+
+    componentWillUnmount: function() {
+        MatrixClientPeg.get().removeListener("deviceVerificationChanged", this._onDeviceVerificationChanged);
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/dialogs/UnknownDeviceDialog.js]: done

File: src/components/views/dialogs/UnknownDeviceDialog.js [hunk 7]
```diff
@@ -83,36 +93,69 @@ function UnknownDeviceList(props) {
...
+        this.props.onSend();
+    },
+
+    _onSendClicked: function() {
+        this.props.onFinished();
+        this.props.onSend();
     },
 
     render: function() {
+        if (this.props.devices === null) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/dialogs/UnknownDeviceDialog.js]: done

File: src/cryptodevices.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Resend from './Resend';
+import sdk from './index';
+import Modal from './Modal';
+import { _t } from './languageHandler';
+
+export function getUnknownDevicesForRoom(matrixClient, room) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/cryptodevices.js]: done

File: src/CallHandler.js (hunk 1 above)
Comment (dbkr) [reply src/CallHandler.js]: This is handled within the js-sdk now - in general we should not be hanging up the call on errors since not all errors are fatal , eg. we may have failed to accept the call but that doesn't mean we (or another client) can't try again.

File: src/CallHandler.js (hunk 2 above)
Comment (dbkr) [reply src/CallHandler.js]: This was only ever used for unknown device errors which are now handled in the general 'error' event (for any other errors sending event, we just did nothing).

File: src/cryptodevices.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/cryptodevices.js]: these guy**s**. `showUnknownDeviceDialogForMessages` and `showUnknownDeviceDialogForCalls` too, please

File: src/CallHandler.js (hunk 3 above)
Comment (richvdh) [reply src/CallHandler.js]: still wondering why this is going away

File: src/cryptodevices.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dbkr) [reply src/cryptodevices.js]: oops, done

File: src/CallHandler.js (hunk 3 above)
Comment (dbkr) [reply src/CallHandler.js]: Ah, because playing the ringback tone ought to indicate that the call is at least probably ringing at the other end, so we set it when the invite has been sent, rather than starting it as soon as we place the call.

Comment (richvdh) [reply src/CallHandler.js]: where does that happen?
